File: ios-native/NeobankNative/Sources/Features/Cards/CardRepository.swift
import Foundation

struct CardRepository: Sendable {
    let client: APIClient

    private struct CardsResponse: Decodable, Sendable {
        let cards: [BankCard]
    }

    func list() async throws -> [BankCard] {
        let response: CardsResponse = try await client.send("/v1/cards")
        return response.cards
    }

    func issue(cardholderName: String, dailyLimit: String?, onlineOnly: Bool?) async throws -> BankCard {
        try await client.send(
            "/v1/cards",
            method: .post,
            body: [
                "cardholder_name": cardholderName,
                "daily_limit": dailyLimit?.isEmpty == false ? dailyLimit : nil,
                "online_only": onlineOnly,
            ]
        )
    }

    func freeze(_ id: String) async throws -> BankCard {
        try await client.send("/v1/cards/\(id)/freeze", method: .post)
    }

    func unfreeze(_ id: String) async throws -> BankCard {
        try await client.send("/v1/cards/\(id)/unfreeze", method: .post)
    }
}

File: android-native/app/src/main/kotlin/dev/horobets/neobanknative/features/cards/CardRepository.kt
package dev.horobets.neobanknative.features.cards

import dev.horobets.neobanknative.core.networking.ApiClient
import dev.horobets.neobanknative.core.networking.HttpMethod
import dev.horobets.neobanknative.core.networking.jsonBodyOf
import kotlinx.serialization.Serializable

class CardRepository(private val client: ApiClient) {
    @Serializable
    private data class CardsResponse(val cards: List<BankCard>)

    suspend fun list(): List<BankCard> = client.send<CardsResponse>("/v1/cards").cards

    suspend fun issue(cardholderName: String, dailyLimit: String?, onlineOnly: Boolean?): BankCard = client.send(
        "/v1/cards",
        method = HttpMethod.POST,
        body = jsonBodyOf(
            "cardholder_name" to cardholderName,
            "daily_limit" to dailyLimit?.ifEmpty { null },
            "online_only" to onlineOnly,
        ),
    )

    suspend fun freeze(id: String): BankCard = client.send("/v1/cards/$id/freeze", method = HttpMethod.POST)

    suspend fun unfreeze(id: String): BankCard = client.send("/v1/cards/$id/unfreeze", method = HttpMethod.POST)
}
